src/parse.rs

```diff
@@ -2552,6 +2552,69 @@ impl<'a> Parser<'a> {
         }
         stack.pop_to(&mut self.tree, 0);
     }
+
+    /// Returns the next event in a pre-order AST walk, along with its
+    /// start and end offset in the source.
+    // TODO: this should probably be destructive. actually remove items from the tree
```

Comment:
There might be scope for both. Non-destructive is one step towards AST access. In any case, since we use the cow pattern so much, I don't think the cloning of strings will be too bad.

Reply:
Those are good points. It's not as clear cut as I thought (I feel like I'm saying this a lot :sweat_smile:) - I will remove the TODO.